bundle/deploy/terraform/state_pull_test.go

```diff
@@ -41,86 +41,104 @@ func statePullTestBundle(t *testing.T) *bundle.Bundle {
-	m := &statePull{
-		identityFiler(mockStateFilerForPull(t, nil, os.ErrNotExist)),
-	}
-
-	ctx := context.Background()
-	b := statePullTestBundle(t)
-	diags := bundle.Apply(ctx, b, m)
-	assert.NoError(t, diags.Error())
-
-	// Confirm that no local state file has been written.
-	_, err := os.Stat(localStateFile(t, ctx, b))
-	assert.ErrorIs(t, err, fs.ErrNotExist)
-}
+func TestStatePullLocal(t *testing.T) {
+	tcases := []struct {
+		name string
 
-func TestStatePullLocalMissingRemotePresent(t *testing.T) {
-	m := &statePull{
-		identityFiler(mockStateFilerForPull(t, map[string]int{"serial": 5}, nil)),
-	}
-
-	ctx := context.Background()
-	b := statePullTestBundle(t)
-	diags := bundle.Apply(ctx, b, m)
-	assert.NoError(t, diags.Error())
+		// remote state before applying the pull mutators
+		remote map[string]any
 
-	// Confirm that the local state file has been updated.
-	localState := readLocalState(t, ctx, b)
-	assert.Equal(t, map[string]int{"serial": 5}, localState)
-}
+		// local state before applying the pull mutators
+		local map[string]any
 
-func TestStatePullLocalStale(t *testing.T) {
-	m := &statePull{
-		identityFiler(mockStateFilerForPull(t, map[string]int{"serial": 5}, nil)),
-	}
-
-	ctx := context.Background()
-	b := statePullTestBundle(t)
-
-	// Write a stale local state file.
-	writeLocalState(t, ctx, b, map[string]int{"serial": 4})
-	diags := bundle.Apply(ctx, b, m)
-	assert.NoError(t, diags.Error())
-
-	// Confirm that the local state file has been updated.
-	localState := readLocalState(t, ctx, b)
-	assert.Equal(t, map[string]int{"serial": 5}, localState)
-}
-
-func TestStatePullLocalEqual(t *testing.T) {
-	m := &statePull{
-		identityFiler(mockStateFilerForPull(t, map[string]int{"serial": 5, "some_other_key": 123}, nil)),
+		// expected local state after applying the pull mutators
+		expected map[string]any
+	}{
+		{
+			name:     "local missing, remote missing",
+			remote:   nil,
+			local:    nil,
+			expected: nil,
+		},
+		{
+			name:     "local missing, remote present",
+			remote:   map[string]any{"serial": 5},
+			local:    nil,
+			expected: map[string]any{"serial": float64(5)},
+		},
+		{
+			name:     "local stale",
+			remote:   map[string]any{"serial": 5},
+			local:    map[string]any{"serial": 4},
+			expected: map[string]any{"serial": float64(5)},
+		},
+		{
+			name:     "local equal",
+			remote:   map[string]any{"serial": 5, "some_other_key": 123},
+			local:    map[string]any{"serial": 5},
+			expected: map[string]any{"serial": float64(5)},
+		},
+		{
+			name:     "local newer",
+			remote:   map[string]any{"serial": 5, "some_other_key": 123},
+			local:    map[string]any{"serial": 6},
+			expected: map[string]any{"serial": float64(6)},
+		},
+		{
+			name:   "remote missing, local present",
+			remote: nil,
+			local:  map[string]any{"serial": 5, "some_other_key": 123},
+			// local state should be invalidated if remote state is missing.
+			expected: nil,
+		},
```

Comment:
These tests without a lineage key depend on two empty strings being equal. They would pass if you rip out the lineage equivalence check. Could you add a check to the mutator that confirms it is not empty?

Reply:
Done